app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -0,0 +1,677 @@
+      const requests = [
+        { method, params },
+        { method, params },
+      ];
+      const mockResults = ['first result', 'second result'];
+
+      await mockingInfuraCommunications(async (comms) => {
+        // The first time a block-cacheable request is made, the block-cache
+        // middleware will request the latest block number through the block
+        // tracker to determine the cache key. Later, the block-ref
+        // middleware will request the latest block number again to resolve
+        // the value of "latest", but the block number is cached once made,
+        // so we only need to mock the request once.
+        comms.mockNextBlockTrackerRequest({ blockNumber: '0x100' });
+        // The block-ref middleware will make the request as specified
+        // except that the block param is replaced with the latest block
+        // number.
+        comms.mockSuccessfulInfuraRpcCall({
+          request: buildRequestWithReplacedBlockParam(
+            requests[0],
+            blockParamIndex,
+            '0x100',
+          ),
+          response: { result: mockResults[0] },
+        });
+        // Note that the block-ref middleware will still allow the original
+        // request to go through.
+        comms.mockSuccessfulInfuraRpcCall({
+          request: requests[0],
+          response: { result: mockResults[0] },
+        });
+
+        const results = await withInfuraClient(({ makeRpcCallsInSeries }) =>
+          makeRpcCallsInSeries(requests),
+        );
+
+        expect(results).toStrictEqual([mockResults[0], mockResults[0]]);
+      });
+    });
+
+    it('hits Infura and does not reuse the result of a previous request if the latest block number was updated since', async () => {
+      const requests = [
+        { method, params },
+        { method, params },
+      ];
+      const mockResults = ['first result', 'second result'];
+
+      await mockingInfuraCommunications(async (comms) => {
+        // Note that we have to mock these requests in a specific order.
+        // The first block tracker request occurs because of the first RPC
+        // request. The second block tracker request, however, does not
+        // occur because of the second RPC request, but rather because we
+        // call `clock.runAll()` below.
+        comms.mockNextBlockTrackerRequest({ blockNumber: '0x100' });
+        // The block-ref middleware will make the request as specified
+        // except that the block param is replaced with the latest block
+        // number.
+        comms.mockSuccessfulInfuraRpcCall({
+          request: buildRequestWithReplacedBlockParam(
+            requests[0],
+            blockParamIndex,
+            '0x100',
+          ),
+          response: { result: mockResults[0] },
+        });
+        // Note that the block-ref middleware will still allow the original
+        // request to go through.
```

Comment:
This is... surprising. How does this work?

It sounds like you're saying the Infura middleware makes _two_ requests for each request, at least for those that use `latest` or have no block tag.

Reply:
I looked this up and found that this is the case: https://github.com/MetaMask/eth-json-rpc-middleware/blob/53c7361944c380e011f5f4ee1e184db746e26d73/src/block-ref.ts#L58

This.... seems wrong. This "child request" with a corrected block number will presumably have its response overwritten by the "real" request.

Reply:
I have confirmed that this is the case. I updated the test to return a bad result for the first of these two semi-duplicated requests, and it was ignored.

Reply:
Looks like this is a bug that was introduced here: https://github.com/MetaMask/eth-json-rpc-middleware/commit/76d676566b0b2be75bc0e952789f4442990a0daf#diff-00d71f9a810b6015201c9aae4ec2d2a5428cc7e2d7745e4267bc36ad58d86bf6R42

Reply:
This was surprising to me as well. I've found a couple of other bugs along the way that I suspect may have been introduced in TypeScript conversions/refactorings.

Reply:
Documented here: https://github.com/MetaMask/eth-json-rpc-middleware/issues/142

I see you raised an issue in that repo already. Was there anything else that hasn't yet been documented?

I do think we should try to fix these issues ASAP. These two seem easy to fix, and the impact might be _very challenging_ to debug if it ever did become an issue in production. It might go a long way towards making our network stack less haunted.

Reply:
Alright. I'll tag this PR as blocked until we can resolve those issues.

The other bugs I found have to do with the fact that the set of methods which take block tags as defined [here](https://github.com/MetaMask/eth-json-rpc-middleware/blob/53c7361944c380e011f5f4ee1e184db746e26d73/src/utils/cache.ts#L53) is incomplete. This means some methods aren't cached properly when they should be. (This list is used indirectly by not only `block-cache` but by a couple of other caching-related middleware as well.)

Reply:
Hmm, OK. That does seem like a bug but not a regression, and not nearly as spooky as the broken retry or the block-ref double-request, both of which were techniques used to guard against inconsistencies resulting from communication with load-balanced servers with different internal state (e.g. from one request to anther, the "latest block" could decrease then increase again as requests were routed to different servers).

Reply:
I didn't mean we should block this PR on fixing those bugs, just that it was important they get fixed. Fixing the bugs would simplify these tests, but we can always simplify them later.

Reply:
Ah, sounds good, okay.